File: 3 course/6 semester/Software development for mobile platforms/films_viewer_kotlin/app/src/main/java/com/example/films_viewer_kotlin/services/AuthService.kt
package com.example.films_viewer_kotlin.services

import com.example.films_viewer_kotlin.models.UserProfile
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthException
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await
import okhttp3.internal.wait

class AuthService {

    private val auth: FirebaseAuth = FirebaseAuth.getInstance()
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()

    val currentUser get() = auth.currentUser

    suspend fun signIn(email: String, password: String): com.google.firebase.auth.FirebaseUser? {
        try {
            val result = auth.signInWithEmailAndPassword(email, password).await()
            // Обновляем поле emailVerified в документе пользователя
            return result.user
        } catch (e: FirebaseAuthException) {
            throw Exception(handleAuthError(e))
        }
    }

    suspend fun signUp(email: String, password: String): com.google.firebase.auth.FirebaseUser? {
        try {
            val result = auth.createUserWithEmailAndPassword(email, password).await()

            val userProfile = UserProfile.empty(result.user!!.uid, result.user!!.email ?: "")
            ProfileService().createInitialProfile(userProfile).await()

            result.user!!.sendEmailVerification().await()
            return result.user
        } catch (e: FirebaseAuthException) {
            throw Exception(handleAuthError(e))
        }
    }

    suspend fun checkEmailVerification() {
        auth.currentUser?.reload()?.await()
    }

    suspend fun resendVerificationEmail() {
        auth.currentUser?.sendEmailVerification()?.await()
    }

    suspend fun deleteUser(password: String) {
        val user = auth.currentUser
        if (user != null) {
            try {
                val email = user.email ?: throw Exception("Email пользователя не найден")

                val credential = EmailAuthProvider.getCredential(email, password)
                user.reauthenticate(credential).await()

                firestore.collection("users").document(user.uid).delete().await()

                val moviesSnapshot = firestore.collection("movies")
                    .whereArrayContains("favoritedBy", user.uid)
                    .get().await()

                for (doc in moviesSnapshot.documents) {
                    val favorites = (doc.get("favoritedBy") as? List<*>)?.toMutableList() ?: mutableListOf()
                    favorites.remove(user.uid)
                    doc.reference.update("favoritedBy", favorites).await()
                }

                user.delete().await()

                signOut();
            } catch (e: FirebaseAuthException) {
                throw Exception(handleAuthError(e))
            } catch (e: Exception) {
                throw Exception("Ошибка при удалении пользователя: ${e.message}")
            }
        }
    }

    suspend fun signOut() {
        auth.signOut()
    }

    private fun handleAuthError(e: FirebaseAuthException): String {
        return when (e.errorCode) {
            "ERROR_REQUIRES_RECENT_LOGIN" -> "Требуется повторный вход"
            "ERROR_UNVERIFIED_EMAIL" -> "Email не подтвержден"
            "ERROR_INVALID_EMAIL" -> "Некорректный email"
            "ERROR_USER_DISABLED" -> "Пользователь заблокирован"
            "ERROR_USER_NOT_FOUND" -> "Пользователь не найден"
            "ERROR_WRONG_PASSWORD" -> "Неверный пароль"
            "ERROR_EMAIL_ALREADY_IN_USE" -> "Email уже зарегистрирован"
            "ERROR_WEAK_PASSWORD" -> "Слишком слабый пароль"
            "ERROR_INVALID_CREDENTIAL" -> "Проверьте логин или пароль"
            else -> "Ошибка аутентификации: ${e.message}"
        }
    }
}

File: 3 course/6 semester/Software development for mobile platforms/films_viewer/lib/Services/AuthService.dart
// auth_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:films_viewer/Services/ProfileService.dart';
import 'package:firebase_auth/firebase_auth.dart';

import '../Models/UserProfile.dart';

class AuthService {
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // Текущий пользователь
  User? get currentUser => _auth.currentUser;

  // Поток состояния авторизации
  Stream<User?> get authStateChanges => _auth.authStateChanges();

  // Вход по email/password
  Future<User?> signIn(String email, String password) async {
    try {
      final credential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );

      return credential.user;
    } on FirebaseAuthException catch (e) {
      throw _handleAuthError(e);
    }
  }

  // Регистрация с отправкой подтверждения
  Future<User?> signUp(String email, String password) async {
    try {
      final credential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );

      var user = UserProfile.empty(credential.user!.uid, email = credential.user!.email!);
      await ProfileService().createInitialProfile(user);

      await credential.user!.sendEmailVerification();
      return credential.user;
    } on FirebaseAuthException catch (e) {
      throw _handleAuthError(e);
    }
  }

  // Проверка подтверждения email
  Future<void> checkEmailVerification() async {
    await _auth.currentUser?.reload();
  }

  // Повторная отправка подтверждения
  Future<void> resendVerificationEmail() async {
    await _auth.currentUser?.sendEmailVerification();
  }

  // Удаление пользователя с повторной аутентификацией
  Future<void> deleteUser(String password) async {
    final user = _auth.currentUser;

    if (user != null) {
      try {
        final email = user.email;
        if (email == null) {
          throw Exception('Email пользователя не найден');
        }

        final credential = EmailAuthProvider.credential(email: email, password: password);
        await user.reauthenticateWithCredential(credential);

        await _firestore.collection('users').doc(user.uid).delete();

        final movies = await _firestore.collection('movies').where('favoritedBy', arrayContains: user.uid).get();
        for (var movie in movies.docs) {
          final favorites = List<String>.from(movie['favoritedBy']);
          favorites.remove(user.uid);
          await movie.reference.update({'favoritedBy': favorites});
        }

        await user.delete();
        await signOut();
      } on FirebaseAuthException catch (e) {
        throw Exception(_handleAuthError(e));
      } catch (e) {
        throw Exception('Ошибка при удалении пользователя: ${e.toString()}');
      }
    }
  }

  // Выход
  Future<void> signOut() async => await _auth.signOut();

  // Обработка ошибок
  String _handleAuthError(FirebaseAuthException e) {
    switch (e.code) {
      case 'requires-recent-login':
        return 'Требуется повторный вход';
      case 'unverified-email':
        return 'Email не подтвержден';
      case 'invalid-email':
        return 'Некорректный email';
      case 'user-disabled':
        return 'Пользователь заблокирован';
      case 'user-not-found':
        return 'Пользователь не найден';
      case 'wrong-password':
        return 'Неверный пароль';
      case 'email-already-in-use':
        return 'Email уже зарегистрирован';
      case 'weak-password':
        return 'Слишком слабый пароль';
      case 'invalid-credential':
        return 'Проверьте логин или пароль';
      default:
        return 'Ошибка аутентификации: ${e.message}';
    }
  }
}